### PR TITLE
Loki: Add support for alerting

### LINKED ISSUE #29175: Loki: Add support for alerting
The  Loki datasource does not support alerting queries

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
This PR adds a simple backend for Loki data source to enable alerting.

I've been running alerts on Loki data source that was set up as Prometheus data sources and with this new support for alerting:
image
image

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/29175

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/tsdb/loki/loki.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+			Start:        start,
+			End:          end,
+			RefId:        queryModel.RefId,
+		})
+	}
+
+	return qs, nil
+}
+
+func parseResponse(value *loghttp.QueryResponse, query *LokiQuery) (*tsdb.QueryResult, error) {
```
Comment (aocenas) [inline pkg/tsdb/loki/loki.go:144]: We don't seem to ever return error here.

File: pkg/tsdb/loki/loki.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+		expr, err := queryModel.Model.Get("expr").String()
+		if err != nil {
+			return nil, err
+		}
+
+		format := queryModel.Model.Get("legendFormat").MustString("")
+
+		start, err := queryContext.TimeRange.ParseFrom()
+		if err != nil {
+			return nil, err
```
Comment (aocenas) [inline pkg/tsdb/loki/loki.go:115]: We could probably wrap these parsing error to be more descriptive. Like `fmt.Errorf("failed to parse From: %v", err)`.

File: pkg/tsdb/loki/loki.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+		}
+		result.Results[query.RefId] = queryResult
+	}
+
+	return result, nil
+}
+
+func formatLegend(metric model.Metric, query *LokiQuery) string {
```
Comment (aocenas) [inline pkg/tsdb/loki/loki.go:85]: Would be nice to have a comment here why do we need this.

File: pkg/tsdb/loki/loki.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+		format := queryModel.Model.Get("legendFormat").MustString("")
+
+		start, err := queryContext.TimeRange.ParseFrom()
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse From: %v", err)
+		}
+
+		end, err := queryContext.TimeRange.ParseTo()
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse From: %v", err)
```
Comment (ivanahuckova) [inline pkg/tsdb/loki/loki.go:121]:
```suggestion
			return nil, fmt.Errorf("failed to parse To: %v", err)
```

File: pkg/tsdb/loki/loki.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+	return string(result)
+}
+
+func parseQuery(dsInfo *models.DataSource, queries []*tsdb.Query, queryContext *tsdb.TsdbQuery) ([]*LokiQuery, error) {
+	qs := []*LokiQuery{}
+	for _, queryModel := range queries {
+		expr, err := queryModel.Model.Get("expr").String()
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse From: %v", err)
```
Comment (ivanahuckova) [inline pkg/tsdb/loki/loki.go:109]:
```suggestion
			return nil, fmt.Errorf("failed to parse Expr: %v", err)
```

File: pkg/tsdb/loki/loki.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+		}
+
+		end, err := queryContext.TimeRange.ParseTo()
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse To: %v", err)
+		}
+
+		dsInterval, err := tsdb.GetIntervalFrom(dsInfo, queryModel.Model, time.Second)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse From: %v", err)
```
Comment (ivanahuckova) [inline pkg/tsdb/loki/loki.go:126]:
```suggestion
			return nil, fmt.Errorf("failed to parse Interval: %v", err)
```

**Review by aocenas — APPROVED** [review]
Look good, nice work

**Comment by loganmc10** [comment]
I just tried this out, am I doing something wrong? Here is my query:

```
sum by (job) (count_over_time({job="eventlogs_dev"} | json[$__interval]))
```

The graph works fine, but this is what I get back from Alerting:

```"tsdb.HandleRequest() error Error response from server: parse error at line 0, col 59: not a valid duration string: "$__interval" (<nil>)"```

Does it not support $__interval? What am I supposed to be using for the time portion in Grafana?

**Comment by ivanahuckova** [comment]
@loganmc10 alerting doesn't support variables. Please use the actual [range vector selector](https://prometheus.io/docs/prometheus/latest/querying/basics/#range-vector-selectors) - e.g. `[10s]`, `[30s]`, `[1m]`, `[5m]`. You can find more on this in here https://grafana.com/docs/loki/latest/logql/#range-vector-aggregation

**Pull request merged**